tooling/greybox_fuzzer/src/coverage.rs

```diff
@@ -0,0 +1,751 @@
+
+/// A map from a particular branch or comparison to its unique index in the raw vector used inside brillig vm
+pub type FeatureToIndexMap = HashMap<Feature, usize>;
+
+/// This file implements the mechanisms for coverage - detection of changes is the execution of the target program
+/// It assists in exploration of the program through testcase mutation by telling the fuzzer whether a new testcase represents previously unexplored
+/// functionality. This in turn allows the fuzzer to add them to the corpus as footholds for further exploration
+///
+/// There are several mechanisms for coverage being used:
+/// 1. Standard branch coverage taken from brillig, the same as with standard, non-zk programs (detects which branch has been taken in an if)
+/// 2. Conditional move coverage from brillig
+/// 3. Novel boolean witness coverage. If ACIR execution was successful, we scan the witness for potential boolean values and detect interesting testcase,
+/// if we discover a boolean witness with a state that hasn't been previously encountered.
+
+/// Mechanism for automated detection of boolean witnesses in the ACIR witness map
+#[derive(Default)]
+pub struct PotentialBoolWitnessList {
+    // Set of witnesses that could be boolean values
+    witness: HashSet<Witness>,
+}
+
+impl From<&WitnessStack<FieldElement>> for PotentialBoolWitnessList {
+    /// Generate a bool witness list by parsing the witnesses in the program
+    fn from(witness_stack: &WitnessStack<FieldElement>) -> Self {
+        let mut witness_set = HashSet::new();
+        // Should be only one function
+        assert!(witness_stack.length() == 1);
+        let first_func_witnesses = witness_stack.peek().unwrap();
+
+        // Look for witnesses that are either 0 or 1
+        for (witness_index, value) in first_func_witnesses.witness.clone().into_iter() {
+            if value == FieldElement::one() || value == FieldElement::zero() {
+                witness_set.insert(witness_index);
+            }
+        }
+        Self { witness: witness_set }
+    }
+}
+
+impl PotentialBoolWitnessList {
+    pub fn new(given_set: HashSet<Witness>) -> Self {
+        Self { witness: given_set }
+    }
+
+    /// Given witnesses from a program, remove non-boolean witnesses from the list
+    pub fn update(&mut self, witness_stack: &WitnessStack<FieldElement>) {
+        assert!(witness_stack.length() == 1);
+        let first_func_witnesses = witness_stack.peek().unwrap();
+        let mut witnesses_for_removal = Vec::new();
+
+        // Go through the list of perceived boolean witnesses
+        for witness_index in self.witness.iter().copied() {
+            let value = first_func_witnesses
+                .witness
+                .get(&witness_index)
+                .expect("There should be a witness in the witness map");
+
+            // Check that the values are zero or 1
+            if *value != FieldElement::zero() && *value != FieldElement::one() {
```

Comment:
For consistency we could use `is_zero`/`is_one` here

Reply:
Done